public/app/core/services/context_srv.ts

```diff
@@ -193,6 +201,80 @@ export class ContextSrv {
     // Hack to reject when user does not have permission
     return ['Reject'];
   }
+
+  // schedules a job to perform token ration in the background
+  private scheduleTokenRotationJob() {
+    // only schedule job if feature toggle is enabled and user is signed in
+    if (config.featureToggles.clientTokenRotation && this.isSignedIn) {
+      // get the time token is going to expire
+      let expires = this.getSessionExpiry();
+
+      // if expires is 0 we run rotation now and reschedule the job
+      // this can happen if user was signed in before upgrade
+      // after a successful rotation the expiry cookie will be present
+      if (expires === 0) {
+        this.rotateToken().then();
+        return;
+      }
+
+      // because this job is scheduled for every tab we have open that shares a session we try
+      // to distribute the scheduling of the job. For now this can be between 1 and 20 seconds
+      const expiresWithDistribution = expires - Math.floor(Math.random() * (20 - 1) + 1);
```

Comment:
Could a [service worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) help having only "one" token rotation job ?

Reply:
That is an interesting idea, have not worked with service workers before but will experiment a bit and see if that could make it easier

Reply:
Will probably explore this in another pr if current logic works fine